sentry_sdk/integrations/falcon.py

```diff
@@ -0,0 +1,162 @@
+        return self.request.cookies
+
+    def raw_data(self):
+        # As request data can only read once we won't make this available
+        # to Sentry.
+        # TODO(jmagnusson): Figure out if there's a way to support this
+        return None
+
+    def form(self):
+        return None  # No such concept in Falcon
+
+    def files(self):
+        return None  # No such concept in Falcon
+
+    def json(self):
+        # We don't touch falcon.Request.media as that can raise an exception
+        # on non-JSON requests.
+        return self.request._media
+
+
+class SentryFalconMiddleware(object):
+    """Captures exceptions in Falcon requests and send to Sentry"""
+
+    def process_request(self, req, resp, *args, **kwargs):
+        hub = Hub.current
+        integration = hub.get_integration(FalconIntegration)
+        if integration is None:
+            return
+
+        with hub.configure_scope() as scope:
+            scope.add_event_processor(
+                _make_request_event_processor(req, integration)
+            )
+
+
+class FalconIntegration(sentry_sdk.integrations.Integration):
+    identifier = "falcon"
+
+    transaction_style = None
+
+    def __init__(self, transaction_style="uri_template"):
+        # type: (str) -> None
+        TRANSACTION_STYLE_VALUES = ("uri_template", "path")
+        if transaction_style not in TRANSACTION_STYLE_VALUES:
+            raise ValueError(
+                "Invalid value for transaction_style: %s (must be in %s)"
+                % (transaction_style, TRANSACTION_STYLE_VALUES)
+            )
+        self.transaction_style = transaction_style
+
+    @staticmethod
+    def setup_once():
+        # type: () -> None
+        _patch_wsgi_app()
+        _patch_handle_exception()
+        _patch_prepare_middleware()
+
+
+def _patch_wsgi_app():
+    original_wsgi_app = falcon.API.__call__
+
+    def sentry_patched_wsgi_app(self, env, start_response):
+        hub = Hub.current
+        integration = hub.get_integration(FalconIntegration)
+        if integration is None:
+            return original_wsgi_app(self, env, start_response)
+
+        sentry_wrapped = SentryWsgiMiddleware(
+            lambda envi, start_resp: original_wsgi_app(self, envi, start_resp)
+        )
+        with hub.push_scope() as scope:
+            scope._name = "falcon"
+            resp = sentry_wrapped(env, start_response)
+
+        return resp
+
+    falcon.API.__call__ = sentry_patched_wsgi_app
+
+
+def _patch_handle_exception():
+    original_handle_exception = falcon.API._handle_exception
+
+    def sentry_patched_handle_exception(self, *args):
+        # NOTE(jmagnusson): falcon 2.0 changed falcon.API._handle_exception
+        # method signature from `(ex, req, resp, params)` to
+        # `(req, resp, ex, params)`
+        if isinstance(args[0], Exception):
+            ex = args[0]
+        else:
+            ex = args[2]
+
+        was_handled = original_handle_exception(self, *args)
```

Comment:
Is this `True` if the user registered a generic 500 error page with Falcon?

What I am getting at is that in many frameworks there is no clear way to know whether the user intended to actually suppress an exception by registering an error handler. I see that you have this testcase where you call `capture_exception` in the errorhandler again, but this means that the integration will not capture anything out of the box for the average application.

For Pyramid we ignore its equivalent of `was_handled` and capture exceptions even if there's an errorhandler. It's unfortunate and people have been complaining about it, but in the face of ambiguity I try to err on the side of sending too much than sending too little.

Reply:
Correct.

Unless an internal error handler is added by the user, all requests will be captured in Sentry. Personally I think it's sensible to not have sentry add any error handlers and let the user decide if the event should still be captured in Sentry. But I can adapt as it's your library and not mine 😉

Reply:
So just let me know if I should change this or not 😊

Reply:
>Personally I think it's sensible to not have sentry add any error handlers and let the user decide if the event should still be captured in Sentry

I assume you would expect the user to decide this by putting/not putting `capture_exception  in their handler?

That's fair but generally we want framework integrations to be useful when the user is just starting to use Sentry, so just calling `init()` should demonstrate the *value* of Sentry to the user. This has been the expectation for every other integration and that's also how we market all of our SDKs. Given that, we have two choices:

* Capture too much, and risk capturing some useless issues
* Capture nothing, which makes the user think our product is broken

The latter results in more support tickets ;)

So yeah please change it. You can still use `before_send` to drop certain exceptions.

Reply:
Thanks, I guess it makes sense in the context of Sentry's core goals. Updated code as suggested.